src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,277 @@
+
+    def _findDevice(self, ports=None, serialnum=None):
+        """
+        Look for a compatible device, requires at least one of the arguments ports and serialnum.
+        ports (str): pattern for the port name
+        serialnum (str): serial number
+        return (str): the name of the port used
+        raises:
+            ValueError: if no device on the ports with the given serial number is found
+        """
+        # At least one of the arguments ports and serialnum must be specified
+        if not ports or serialnum:
+            raise ValueError("At least one of the arguments 'ports' and 'serialnum' must be specified.")
+
+        # For debugging purpose
+        if ports == "/dev/fake":
+            return ports
+
+        # If no ports specified, check all available ports
+        if ports:
+            names = glob.glob(ports)
+        else:
+            names = serial.tools.list_ports.comports()  # search all serial ports
+
+        # Look for serial number if available, otherwise make sure only one port matches the port pattern.
+        if serialnum:
+            for port in names:
+                if serialnum in port.hwid:
+                    self._serial = self._openSerialPort(port)
+                    return port  # Found it!
+            else:
+                raise ValueError("Couldn't find device with serial number %s in list of ports %s." % (serialnum, names))
+        else:
+            if len(names) == 1:
+                port = names[0]
+                self._serial = self._openSerialPort(port)
+                return port
+            elif len(names) > 1:
+                raise ValueError("Multiple ports detected. Please specify a serial number.")
+            else:
+                raise ValueError("Couldn't find port with pattern '%s'." % ports)
+
+    def _openSerialPort(self, port):
+        if self._port == "/dev/fake":
+            return BeamControllerSimulator()
+        else:
+            return ModbusSerialClient(method='rtu', port=port,
+                                      baudrate=BAUDRATE, timeout=TIMEOUT,
+                                      stopbits=STOPBITS, parity=PARITY,
+                                      bytesize=BYTESIZE)
+
+    def _setShift(self, value):
+        """
+        :param value (float, float): x, y shift from the center (in m)
+        """
+        gain = self._metadata.get(model.MD_CALIB_SCALE)
+        if not gain:
+            raise ValueError("No ampere-to-meter conversion factor specified. Please run the alignment procedure "
+                             "first.")
+        offset = self._metadata.get(model.MD_CALIB_TRANSLATION)
+        if not offset:
+            logging.warning("No deflection offset specified, assuming 0 amps.")
+            offset = (0, 0)
+
+        rotation = self._metadata.get(model.MD_CALIB_ROTATION)
+        if not offset:
+            logging.warning("No rotation specified, assuming 0 rad.")
+            rotation = 0
```

Comment:
You could initialise this metadata to the default value, so that if nothing is done, it'd be visible in the "odemis-cli" (ie .getMetadata). 
The fact that you cannot set the value if there is no calibration is a little bit arguable. One issue I imagine is that if the VA is set as persistent (which it shouldn't) in the backend, then the whole component cannot start (because VAs are set before metadata). That said, maybe that's what should happen, as it's kind of meaningly. It's also weird that there is a value, but you cannot set it. Maybe don't raise an error if someone set the same value? I do agree it's also weird to attempt changing the value without any calibration.
Hopefully, in reality, we'll have a pretty good value by default in the YAML file, and this code will never be triggered anyway.

Reply:
How about something like this:
```
        gain = self._metadata.get(model.MD_CALIB_SCALE)
        if not gain:
            gain = DEFAULT_GAIN
            logging.error("No gain specified, assuming default of %s." % gain +
                          "Calibration is required to set shift in a meaningful way.")
            self.updateMetadata({model.MD_CALIB_SCALE: gain})
```

Reply:
ok, without the last line, that doesn't make sense (otherwise the error will not be triggered the next time)